new actions, code style consistency, async await, disabled actions and bugfixes

dbuezas opened the pull request:
The options html is now prettier. Hope you like it :)

Commit:
added Reopen Closed Tab suggestion

Commit:
using async await

dbuezas commented:
<img width="690" alt="screen shot 2017-08-05 at 19 17 48" src="https://user-images.githubusercontent.com/777196/28997371-dcccfe28-7a12-11e7-8d65-e5d2852649a3.png">

dbuezas commented:
everything according to existing style :)
I also replaced all double quotes to single quotes and ported all actions to await async. That saves 65 lines of code, a lot of columns, reduces nesting and improves readability.
the only fat arrow functions now are one liners inside .filter or .map. Those return a value so there it is worth it :)
Hope you like it. I love this extension by the way. These new javascript syntax sugar could also attract more people to add actions, since I think they are now way easier to understand and follow.

BTW, the fuzzy search module seems to sort in a sort of wacky way, it doesn't care if you have a full word match.

Commit:
added "open commander options" action

ssundarraj commented:
> Hope you like it. I love this extension by the way. These new javascript syntax sugar could also attract more people to add actions, since I think they are now way easier to understand and follow.

Thanks. This is great! I should probably get around to porting the rest of the syntax to use ES6 some time soon :)
> BTW, the fuzzy search module seems to sort in a sort of wacky way, it doesn't care if you have a full word match.

That's odd. Can you please file a bug with a screenshot if you have one handy?

ssundarraj commented:
This is what I get when I build the extension and test it. Seems broken to me. Maybe you're missing some changes?

![screen shot 2017-08-06 at 12 33 29 am](https://user-images.githubusercontent.com/4509724/29001416-f18cdc1c-7a3e-11e7-97cb-2adf09519793.png)

dbuezas commented:
> Thanks. This is great! I should probably get around to porting the rest of the syntax to use ES6 some time soon :)

Cool! I can definitely recommend to use es-lint, there are some non declared variables that may lead to bugs and linting tells you about that :). Also it would help defining a coding style and having everybody respect it, plus it also gives recommendations about best practices, etc. Can help setting that up if you wish.
It would also be great to require other files instead of using globals, but at the current size of the project it is still manageable.

> That's odd. Can you please file a bug with a screenshot if you have one handy?

<img width="409" alt="screen shot 2017-08-06 at 11 31 09" src="https://user-images.githubusercontent.com/777196/29002114-8625feec-7a9b-11e7-8434-e5994a87e055.png">
in this case I would expect the algorithm to prioritise `Switch to: [....] Youtube`

dbuezas commented:
The bug you found is fixed now 🎉 
I accidentally changed the format in which custom actions were stored and didn't notice it myself because I "fixed" it in the `populateSuggestionList` function instead of in the `addCommand`function 😅 

While I was debugging I got stuck because the stored custom action's code wasn't valid, so I added a couple of small features:

1. Don't store custom actions that cannot be 'eval'ed.
<img width="253" alt="screen shot 2017-08-06 at 11 41 03" src="https://user-images.githubusercontent.com/777196/29002139-2e593e30-7a9c-11e7-966f-b292a1d10d51.png">

2. Show an error message in the popup if an action throws:
<img width="415" alt="screen shot 2017-08-06 at 11 29 26" src="https://user-images.githubusercontent.com/777196/29002156-79ed6fce-7a9c-11e7-8b09-75b2b7cfe01d.png">
It also points out to open the popup's DevTools and console.errors the function and the exception so users don't get lost while debugging. This also applies to default actions, so it should help extension devs also (I got lost a bit at first).
<img width="490" alt="screen shot 2017-08-06 at 11 51 48" src="https://user-images.githubusercontent.com/777196/29002211-aae07b48-7a9d-11e7-9ff5-3a872c2cec6d.png">

3. I added a link to the chrome api to the user custom actions comments:
<img width="560" alt="screen shot 2017-08-06 at 11 49 25" src="https://user-images.githubusercontent.com/777196/29002194-5a5b1f84-7a9d-11e7-8840-12ccd5c5fd01.png">
I wished I had that at hand yesterday when I started adding the "reopen closed tab" action :)


4. Open Commander Options from the popup:
<img width="409" alt="screen shot 2017-08-06 at 11 50 23" src="https://user-images.githubusercontent.com/777196/29002201-77189b2e-7a9d-11e7-97b5-97cbc2df239c.png">

dbuezas commented:
the new fuzzy library works great! 
I simplified stuff a bit and reduced code size another 20 lines in main.js.
also now searches can be added in the actions.js like this:
```{
    text: 'Define in Dictionary.com',
    action: triggerSearch('def'),
    triggers: ['def', 'define', 'dictionary'],
    queryToUrl: (q) => `http://dictionary.reference.com/browse/${q}`,
  }
```
and can also be disabled in the options since they are handled like normal actions now

dbuezas commented:
I have a gizillion tabs open and switch youtube now only gives the correct result 🎉 
<img width="411" alt="screen shot 2017-08-06 at 14 26 35" src="https://user-images.githubusercontent.com/777196/29003243-563c7aae-7ab3-11e7-8b38-a4128cd30728.png">

ssundarraj commented:
The bug still exists for me. I suspect it's because some of the changes aren't compatible with the grunt config. When you use the extension are you using the distribution version of it?

If not, can you run `grunt` and load the extension from the 'dist' directory and see if it works? 

<img width="626" alt="screen shot 2017-08-07 at 11 31 55 pm" src="https://user-images.githubusercontent.com/4509724/29059149-fe91bb90-7bc8-11e7-93be-479d52815c49.png">

ssundarraj commented:
I think you'll need to update L29 of the gruntfile.

dbuezas commented:
yep, never tried to build the extension 😅.
This should fix it.
I added a simple regex to replace the css to their .min.css and another one to replace everything between
`<!-- <jsminify dest="some.file.name"> -->`
and
`<!-- </jsminify> -->`
with a script tag with src = `<whatever was in dest>`

one still has to add the required js files in the Gruntfile.js but at least the replace term is easier (I think)

ssundarraj commented on Gruntfile.js:
```diff
@@ -24,10 +34,16 @@ module.exports = function(grunt) {
-            content = content.replace("js/options.js", "options.min.js");
-            return content.replace("<script src=\"js/vendor/fuse.min.js\"></script>\n    <script src=\"js/actions.js\"></script>\n    <script src=\"js/main.js\"></script>", "<script src=\"built.min.js\"></script>").replace("css/main.css", "main.min.css");
-          }
+                return content
```
Indentation looks off here

dbuezas replied:
it was off before, old lines 27 to 29 are inside the process function.
regarding the `.replace` lines, I'd rather have them with the same identation as the return statement, what is your preference?

dbuezas replied:
optn1:
```
process: function (content, srcpath){
    return content
      .replace(
        /href="css\/([^"]+?)\.css"/g,
        'href="$1.min.css"'
      )
      .replace(
        /<!-- <jsminify dest="(.+)"> -->[\S\s]+?<!-- <\/jsminify> -->/g,
        '<script src="$1"></script>'
      );
}
```
optn2:
```
process: function (content, srcpath){
    return content
    .replace(
      /href="css\/([^"]+?)\.css"/g,
      'href="$1.min.css"'
    )
    .replace(
      /<!-- <jsminify dest="(.+)"> -->[\S\s]+?<!-- <\/jsminify> -->/g,
      '<script src="$1"></script>'
    );
}
```
optn3:
```
process: function (content, srcpath){
    return content.replace(
      /href="css\/([^"]+?)\.css"/g,
      'href="$1.min.css"'
    ).replace(
      /<!-- <jsminify dest="(.+)"> -->[\S\s]+?<!-- <\/jsminify> -->/g,
      '<script src="$1"></script>'
    );
}
```

ssundarraj replied:
Oh my bad. Please leave it as is.

ssundarraj commented on src/js/actions.js:
```diff
@@ -1,312 +1,262 @@
-    }
-}
-
+chromeP = new ChromePromise();
```
Please rename chromeP -> chromePromise

ssundarraj commented on src/js/main.js:
```diff
@@ -1,122 +1,111 @@
-        reScroll();
-        return false;
+  switch (e.which){
+    case (40):{// down
```
nit: add space before the comment

ssundarraj commented on src/js/main.js:
```diff
@@ -125,42 +114,75 @@ function handleMouseover(e){
+            await suggestion.action();
+          } catch (e) {
+            document.body.innerHTML = (
+`Error executing action [${escapeHtml(suggestion.text)}]:
```
Can this be indented?

dbuezas replied:
you're right, it is HTML so it doesn't matter

ssundarraj reviewed (commented):
Looks good minus some minor changes. Thanks!

Can you also clean up the commit log so it looks neater? It would be great if you can squash commits that are part of the same function into a single commit but if that's not possible, I'd settle for one monolithic commit with better descriptions.

Commit:
adds highlighting of fuzzy search

Commit:
fixes popup staying tiny some times

Commit:
adds attach/deatach tab from window actions

Commit:
adds ability to remove suggestions (solves #10) and prettier user actions

Commit:
homogenizes code style and use of async await

Commit:
fixes user actions and improves usability of editor

Commit:
replaced fuzzy search algorithm and improoved handling of search engines feature

Commit:
fixed grunt and codestyle

dbuezas commented:
this took way too much effort but commits are now squashed and I learned some more git stuff :P

ssundarraj commented:
Thanks David!

dbuezas commented:
🎉 :)

dbuezas commented:
next PRs should be more orderly
